Make internal classloaders parallel capable

eatkins opened the pull request:
Because classloading with a reverse lookup classloader is not strictly
hierarchical, it is not safe to treat the BottomClassLoader as parallel
compatibile. To address this, I rework things so that the
BottomClassLoader actually synchronizes in the loadClass implementation.
I also updated the java serialization scripted test in a way that can
trigger a deadlock without synchronization.

This had no measurable performance impact when I reran
https://github.com/eatkins/scala-build-watch-performance

eatkins commented:
May fix #5116.

eed3si9n commented:
To test the concurrency of `test` task, it might be interesting to add a bunch of long running tests that holds on to various loaded things. For example, in the originally reported issue it's trying to initialize `cats.data.NonEmptyList` companion object.

eatkins commented:
That could be interesting. This is actually a pretty good test because it actually is failing quite reliably. I didn't test it extensively enough before pushing.

I'm starting to think that java serialization is maybe a lost cause and we should just throw a very specific exception that fails the test and explains that either the code needs to be refactored to not use java serialization or change the layering strategy. If everything is single-threaded, the reverse lookup seems ok, but I'm worried that it's going to take forever to get it right in the presence of concurrency. You already lose the benefit of turbo mode because we have to invalidate the dependency layer when we use java serialization so it's probably better to just fail hard.

eed3si9n commented:
Is #5116 related to serialization?

eatkins committed:
Use java to implement XMain classloaders

These classloaders which are created if sbt is launched with a legacy
launcher (or one that doesn't follow the current classloading hierarchy
convention), were implemented in scala, but that meant that they were
not parallel capable. I fix that by moving the implementations to java.
I also move the static method that creates a MetaBuildLoader into the
java class.

eed3si9n commented:
Could we add the repro in https://github.com/sbt/sbt/issues/5116#issuecomment-535184386?

eatkins commented:
The repro is the java-serialization scripted test.

eatkins commented on sbt/src/sbt-test/classloader-cache/java-serialization/descendant/src/test/scala/test/ReflectionTest.scala:
```diff
@@ -1,12 +1,32 @@
+        initLatch.countDown()
+        initLatch.await(5, TimeUnit.SECONDS)
+        val className = if (i % 2 == 0) "test.Foo" else "test.Bar"
+        loader.loadClass(className)
```
`test.Foo` extends `test.Bar` so we should end up concurrently trying to load `test.Bar`. This can cause deadlock if the classloaders are not parallel capable.

eed3si9n commented on main/src/main/java/sbt/internal/BottomClassLoader.java:
```diff
@@ -0,0 +1,101 @@
+    ClassLoader.registerAsParallelCapable();
+  }
+
+  final Class<?> reverseLoad(final String name, final boolean resolve)
```
What does reverseLoad mean?

eed3si9n replied:
Since loading is `String => Class`, a reverse would be `Class => String`.

Is `loadOnceForParent(String, Boolean): Class` what's going on?

eatkins replied:
I chose reverseLoad to mean that we were traversing the classloading hierarchy in the _reverse_ direction. Normally a child loads a class from it’s parent but here the parent tries to load a class from a child.

eed3si9n replied:
Right, and the `BoomClassLoader` is the child who gets sacrificed by the parent right? So I think `loadOnce` or `loadForParent` might be a better name with comment saying it gets called by the parent.

Also instead of reversing, would it make sense to have two parents, like a primary parent, and secondary parent? If the primary fails for whatever reason lazily create a flat `ClassLoader`?

eatkins replied:
Instead of renaming `reverseLoad`, I just removed it and made `findClass` public in `BottomClassLoader`.

eatkins replied:
> Also instead of reversing, would it make sense to have two parents, like a primary parent, and secondary parent? If the primary fails for whatever reason lazily create a flat ClassLoader?

I don't think I understand this idea. Is the idea that we'd create a dual loader where the primary classloader has the dependency classpath and the secondary loader has the full classpath?

eed3si9n replied:
Yea. So in that world, there's no traversing back to child, it would just fall back a flat secondary parent.

eatkins replied:
~~That would not work because the classes loaded by the secondary parent cannot be used by the child.~~
The reason for fall back to the child is specifically for when java reflection is used in the parent layer to access classes that are only visible to the child. Loading those classes in alternate secondary parent loader creates an alternate classloading hierarchy which would be incompatible with the normal hierarchy and would almost surely cause frequent incompatible class exceptions.

eatkins replied:
What I've been trying to do is isolate things so that there is at most one classloader that can load each class on the classpath (this could be violated if multipe jars define the same class). Dual loaders make it possible for multiple classloaders to load the same class which makes it much easier for something to go wrong.

eatkins replied:
To make it concrete, the java-serialization scripted test defines an object that looks like this:
```
object Reflection {
  def roundTrip[A](a: A): A = {
    val baos = new ByteArrayOutputStream()
    val oos = new ObjectOutputStream(baos)
    oos.writeObject(a)
    oos.close()
    val bais = new ByteArrayInputStream(baos.toByteArray())
    val ois = new ObjectInputStream(bais)
    try ois.readObject().asInstanceOf[A]
    finally ois.close()
  }
}
```
The call to `ois.readObject` will eventually lead to a call to `Class.forName` with the class name read from the object stream. Suppose that we invoke this method with a class `A` that is defined in the project classpath. If there were a secondary flatloader that were used to load `A`, then the `asInstanceOf[A]` would cause a `ClassCastException`. By looking back into the child, however, we load the same version of `A` that was passed in as input.

eed3si9n replied:
> Loading those classes in alternate secondary parent loader creates an alternate classloading hierarchy which would be incompatible with the normal hierarchy and would almost surely cause frequent incompatible class exceptions.

Yea. That makes sense.

eatkins committed:
Make all test and run classloaders parallel capable

A number of users were reporting issues with deadlocking when using
1.3.2: https://github.com/sbt/sbt/issues/5116. This seems to be because
most of the sbt created classloaders were not actually parallel capable.
In order for a classloader to be registered as a parallel capable, ALL
of the parent classes except for object in the class hierarchy must be
registered as a parallel capable:
https://docs.oracle.com/javase/8/docs/api/java/lang/ClassLoader.html#registerAsParallelCapable--.
If a classloader is not registered as parallel capable, then a global
lock will be used internally for classloading and this can lead to deadlock.

It is impossible to register a scala 2 classloader as parallel capable
so I ported all of the classloaders to java.

This commit updates the java-serialization scripted test. Prior to the
port, the new version of the test would more or less always deadlock.
After this change, I haven't been able to reproduce a deadlock.

This had no significant performance impact when I reran
https://github.com/eatkins/scala-build-watch-performance

eatkins committed:
Bump travis java 8 version

eed3si9n reviewed (dismissed):
LGTM pending CI

eatkins commented:
Passing now.
